src/v/cloud_topics/level_one/metastore/frontend.cc

```diff
@@ -0,0 +1,409 @@
+  , _partition_mgr(partition_mgr)
+  , _topics_frontend(topics_frontend)
+  , _metadata(metadata)
+  , _leaders(leaders)
+  , _shard_table(shards)
+  , _connection_cache(connections) {}
+
+ss::future<> frontend::stop() { return _gate.close(); }
+
+ss::future<bool> frontend::ensure_topic_exists() {
+    // todo: make these configurable.
+    static constexpr int16_t default_replication_factor = 3;
+    static constexpr int32_t default_metastore_partitions = 3;
+
+    const auto& metadata = _metadata->local();
+    if (metadata.get_topic_metadata_ref(model::l1_metastore_nt)) {
+        co_return true;
+    }
+    auto replication_factor = default_replication_factor;
+    if (replication_factor > static_cast<int16_t>(metadata.node_count())) {
+        replication_factor = 1;
+    }
+
+    cluster::topic_configuration topic{
+      model::kafka_internal_namespace,
+      model::l1_metastore_topic,
+      default_metastore_partitions,
+      replication_factor};
+
+    topic.properties.compression = model::compression::none;
+    // todo: fix this by implementing on demand raft
+    // snapshots.
+    topic.properties.cleanup_policy_bitflags
+      = model::cleanup_policy_bitflags::none;
+    topic.properties.retention_bytes = tristate<size_t>();
+    topic.properties.retention_local_target_bytes = tristate<size_t>();
+    topic.properties.retention_duration = tristate<std::chrono::milliseconds>();
+    topic.properties.retention_local_target_ms
+      = tristate<std::chrono::milliseconds>();
+
+    try {
+        auto res = co_await _topics_frontend->local().autocreate_topics(
+          {std::move(topic)},
+          config::shard_local_cfg().create_topic_timeout_ms());
+        vassert(
+          res.size() == 1,
+          "Incorrect result when creating {}, expected 1 response, "
+          "got: {}",
+          model::l1_metastore_topic,
+          res.size());
+        if (
+          res[0].ec != cluster::errc::success
+          && res[0].ec != cluster::errc::topic_already_exists) {
+            vlog(
+              cd_log.warn,
+              "can not create topic: {} - error: {}",
+              model::l1_metastore_topic,
+              cluster::make_error_code(res[0].ec).message());
+            co_return false;
+        }
+        co_return true;
+    } catch (const std::exception_ptr& e) {
+        vlog(
+          cd_log.warn,
+          "can not create topic: {} - error: {}",
+          model::l1_metastore_topic,
+          e);
+        co_return false;
+    }
+}
+
+std::optional<model::partition_id>
+frontend::metastore_partition(const model::topic_id_partition& tp) const {
+    const auto md = _metadata->local().get_topic_metadata_ref(
+      model::l1_metastore_nt);
+    if (!md) {
+        return std::nullopt;
+    }
+    iobuf temp;
+    write(temp, tp);
+    auto bytes = iobuf_to_bytes(temp);
+    auto partition = murmur2(bytes.data(), bytes.size())
+                     % md->get().get_configuration().partition_count;
```

Comment:
we'll want to prevent partition count from being alterned on this topic. we might also want to do something defensive where we "register" new topic/partitions in a metastore partition and then verify that we aren't getting any misdirected RPCs.

Reply:
I wonder if we should just explicitly hardcode to partition 0 for now 🤔

Reply:
I don't think we should handle changing number of partitions right now. Ideally we can tackle that when we tackle the broader load balancing challege. But you're right andrew, if we really needed to do it, we could track the history of the partition count changes. then the partition function would determine which partition count was used when a partition was created (or something along these lines).

Reply:
I don't think there's harm in having multiple partitions right now -- it's a hopefully painless, assuming no migrations, thing to keep in, while giving us a real-ish idea of how domains will eventually work. I can change it if you feel strongly about it though.

Migrations OTOH, I think we should punt on, until maybe even after the first cut

Reply:
I don't feel strongly about anything other than punting on migrations :) I'm fine with allowing multiple partitions to start, but just needs clear expectations it can't be changed.